x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -51,83 +55,178 @@
+                    listener.onFailure(
+                        new ElasticsearchStatusException(Messages.INFERENCE_FAILED_TO_STORE_MODEL,
+                            RestStatus.INTERNAL_SERVER_ERROR,
+                            e,
+                            trainedModelConfig.getModelId()));
+                }
+            }
+        );
+
+        ActionListener<IndexResponse> putConfigListener = ActionListener.wrap(
+            r -> {
+                if (trainedModelConfig.getDefinition() != null) {
+                    indexObject(TrainedModelDefinition.docId(trainedModelConfig.getModelId()),
+                        trainedModelConfig.getDefinition(),
+                        putDefinitionListener);
+                } else {
+                    listener.onResponse(true);
+                }
+            },
+            e -> {
+                logger.error(new ParameterizedMessage(
+                    "[{}] failed to store trained model for inference", trainedModelConfig.getModelId()), e);
+                if (ExceptionsHelper.unwrapCause(e) instanceof VersionConflictEngineException) {
+                    listener.onFailure(new ResourceAlreadyExistsException(
+                        Messages.getMessage(Messages.INFERENCE_TRAINED_MODEL_EXISTS, trainedModelConfig.getModelId())));
+                } else {
+                    listener.onFailure(
+                        new ElasticsearchStatusException(Messages.INFERENCE_FAILED_TO_STORE_MODEL,
+                            RestStatus.INTERNAL_SERVER_ERROR,
+                            e,
+                            trainedModelConfig.getModelId()));
+                }
+            }
+        );
+
+        indexObject(trainedModelConfig.getModelId(), trainedModelConfig, putConfigListener);
     }
 
-    public void getTrainedModel(String modelId, ActionListener<TrainedModelConfig> listener) {
+    public void getTrainedModel(final String modelId, final boolean includeDefinition, final ActionListener<TrainedModelConfig> listener) {
+
         QueryBuilder queryBuilder = QueryBuilders.constantScoreQuery(QueryBuilders
             .idsQuery()
             .addIds(modelId));
-        SearchRequest searchRequest = client.prepareSearch(InferenceIndexConstants.INDEX_PATTERN)
-            .setQuery(queryBuilder)
-            // use sort to get the last
-            .addSort("_index", SortOrder.DESC)
-            .setSize(1)
-            .request();
-
-        executeAsyncWithOrigin(client, ML_ORIGIN, SearchAction.INSTANCE, searchRequest,
-            ActionListener.wrap(
-                searchResponse -> {
-                    if (searchResponse.getHits().getHits().length == 0) {
+        MultiSearchRequestBuilder multiSearchRequestBuilder = client.prepareMultiSearch()
+            .add(client.prepareSearch(InferenceIndexConstants.INDEX_PATTERN)
+                .setQuery(queryBuilder)
+                // use sort to get the last
+                .addSort("_index", SortOrder.DESC)
```

Comment:
Unrelated to this change but why are we sorting on `_index` here instead of `create_time`?

Reply:
I think we should prefer the latest index version regardless of when the model is created. If we support updating, an updated model may go into the new index, but have an older create time of another model.

This is an open question for sure, but this is consistent with how we handle transforms.